swift/ql/lib/codeql/swift/frameworks/StandardLibrary/FilePath.qll

```diff
@@ -15,6 +19,76 @@ class FilePath extends StructDecl {
+        ";FilePath;true;init(platformString:);;;Argument[0];ReturnValue;taint",
+        ";FilePath;true;init(root:_:);;;Argument[0..1];ReturnValue;taint",
+        ";FilePath;true;init(root:components:);;;Argument[0..1];ReturnValue;taint",
+        ";FilePath;true;encode(to:);;;Argument[-1];Argument[0];taint",
+        ";FilePath;true;withCString(_:);;;Argument[-1];Argument[0].Parameter[0];taint",
+        ";FilePath;true;withPlatformString(_:);;;Argument[-1];Argument[0].Parameter[0];taint",
+        ";FilePath;true;append(_:);;;Argument[0];Argument[-1];taint",
+        ";FilePath;true;appending(_:);;;Argument[-1..0];ReturnValue;taint",
+        ";FilePath;true;lexicallyNormalized();;;Argument[-1];ReturnValue;taint",
+        ";FilePath;true;lexicallyResolving(_:);;;Argument[-1..0];ReturnValue;taint",
+        ";FilePath;true;push(_:);;;Argument[0];Argument[-1];taint",
+        ";FilePath;true;pushing(_:);;;Argument[-1..0];ReturnValue;taint",
+        ";FilePath;true;removingLastComponent();;;Argument[-1];ReturnValue;taint",
+        ";FilePath;true;removingRoot();;;Argument[-1];ReturnValue;taint",
+        ";FilePath.Component;true;init(_:);;;Argument[0];ReturnValue;taint",
+        ";FilePath.Component;true;init(platformString:);;;Argument[0];ReturnValue;taint",
+        ";FilePath.Component;true;withPlatformString(_:);;;Argument[-1];Argument[0].Parameter[0];taint",
+        ";FilePath.Root;true;init(_:);;;Argument[0];ReturnValue;taint",
+        ";FilePath.Root;true;init(platformString:);;;Argument[0];ReturnValue;taint",
+        ";FilePath.Root;true;withPlatformString(_:);;;Argument[-1];Argument[0].Parameter[0];taint"
+      ]
+  }
+}
+
+/**
+ * A content implying that, if a `FilePath` is tainted, certain fields are also
+ * tainted.
+ */
+private class FilePathFieldsInheritTaint extends TaintInheritingContent,
+  DataFlow::Content::FieldContent
+{
+  FilePathFieldsInheritTaint() {
+    exists(FieldDecl f | this.getField() = f |
+      (
+        f.getEnclosingDecl().(NominalTypeDecl) instanceof FilePath or
+        f.getEnclosingDecl().(ExtensionDecl).getExtendedTypeDecl() instanceof FilePath
+      ) and
+      f.getName() =
+        [
+          "description", "debugDescription", "components", "extension", "lastComponent", "root",
+          "stem", "string"
+        ]
+    )
+  }
+}
+
+/**
+ * A content implying that, if a `FilePath.Component` or `FilePath.Root` is tainted, certain fields
+ * are also tainted.
+ */
+private class FilePathComponentFieldsInheritTaint extends TaintInheritingContent,
+  DataFlow::Content::FieldContent
+{
+  FilePathComponentFieldsInheritTaint() {
+    exists(FieldDecl f | this.getField() = f |
+      (
+        f.getEnclosingDecl().(NominalTypeDecl).getFullName() =
+          ["FilePath.Component", "FilePath.Root"] or
+        f.getEnclosingDecl().(ExtensionDecl).getExtendedTypeDecl().getName() =
+          ["FilePath.Component", "FilePath.Root"]
```

Comment:
Could we not use `asNominalTypeDecl` to collapse these two disjuncts into one? Or is it crucial that we separate them into two cases because we need `getFullName` in one case and `getName` in the other?

Reply:
I'm actually looking into whether it's possible to combine the disjuncts now - but it's an issue that affects multiple models, not just this one, so I'll probably fix in a separate PR (if they can be combined).

The `getFullName` vs `getName` is a typo, I've just pushed a fix for that.

Reply:
PR for the excessive disjuncts here: https://github.com/github/codeql/pull/13223

(we already had a solution for this pattern, but had failed to use it widely)